scipy/spatial/_hausdorff.pyx

```diff
@@ -0,0 +1,63 @@
+"""
+Directed Hausdorff Code
+
+.. versionadded:: 0.19.0
+
+"""
+#
+# Copyright (C)  Tyler Reddy and Richard Gowers, 2016
+#
+# Distributed under the same BSD license as Scipy.
+#
+
+import numpy as np
+cimport numpy as np
+cimport cython
+from libc.math cimport sqrt
+
+__all__ = ['directed_hausdorff']
+
+@cython.boundscheck(False)
+def directed_hausdorff(double[:,::1] ar1, double[:,::1] ar2):
+
+    cdef double cmax, cmin, d
+    cdef int break_occurred
+    cdef int N1 = ar1.shape[0]
+    cdef int N2 = ar2.shape[0]
+    cdef int data_dims = ar1.shape[1]
+    cdef unsigned int i, j, k, i_store, j_store, i_ret, j_ret
+    cdef long[:] resort1, resort2
+
+    # shuffling the points in each array generally increases the likelihood of
+    # an advantageous break in the inner search loop and never decreases the
+    # performance of the algorithm
+    np.random.seed(123)
```

Comment:
As noted above, I'm open to providing the user with the option to give the seed integer as a keyword argument. I haven't seen concrete proof of any scenario where the directed Hausdorff distance proper would be incorrect as a result of the seed -- that should be impossible -- it should only affect the performance of the algorithm. 

If you mean that setting the random state globally in a way that is hidden to the user is potentially risky, I can agree with that, but it would still never give the wrong result for the function proper. If your suggestion above for shuffling with no seed can work in the context of the algorithm, that may avoid the issue of impacting external code with the (hidden) seed fixing. I'm not sure if there is a nice way to "release" the seed back to using machine time just before the function exits--that might also work.

In any case, I've already highlighted some caveats related to the use of a fixed seed in my notes above, and those are directly related to the algorithm proper, so I wasn't really that keen to fix the seed in the first place (this is a first draft of added functionality).